File: entity-meta-generator-lib/src/main/java/com/loudsight/meta/serialization/transform/CustomEntityTransform.java
package com.loudsight.meta.serialization.transform;

import com.loudsight.meta.Meta;
import com.loudsight.meta.MetaRepository;
import com.loudsight.meta.serialization.EntityTransform;
import com.loudsight.meta.serialization.EntityType;

import com.loudsight.useful.helper.logging.LoggingHelper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.IdentityHashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Set;

public class CustomEntityTransform extends EntityTransform<Object> { 
    private static final LoggingHelper logger = LoggingHelper.wrap(CustomEntityTransform.class);
    
    private static final int MAX_SERIALIZATION_DEPTH = 100;
    
    private static final ThreadLocal<Set<Object>> SERIALIZATION_STACK = 
            ThreadLocal.withInitial(() -> Collections.newSetFromMap(new IdentityHashMap<>()));
    
    private static final ThreadLocal<Integer> SERIALIZATION_DEPTH = 
            ThreadLocal.withInitial(() -> 0);

    private static class CustomEntityTransformHolder {
        private static final CustomEntityTransform INSTANCE = new CustomEntityTransform();
    }
    // global access point
    public static CustomEntityTransform getInstance() {
        return CustomEntityTransformHolder.INSTANCE;
    }
    private CustomEntityTransform() {
        super(EntityType.CUSTOM, Object.class);
    }


    @Override public void serializeEntity( Object entity, List<Byte> bytes) {
        serializeEntityX(entity, bytes);
    }

    private <T> void serializeEntityX(T entity, List<Byte> bytes) {
        Set<Object> stack = SERIALIZATION_STACK.get();
        int depth = SERIALIZATION_DEPTH.get();
        
        try {
            if (stack.contains(entity)) {
                String errorMsg = String.format(
                        "Circular reference detected while serializing %s (identity=%d). " +
                        "Serialization stack contains %d objects. Entity: %s",
                        entity.getClass().getName(),
                        System.identityHashCode(entity),
                        stack.size(),
                        safeToString(entity));
                logger.logError(errorMsg);
                throw new IllegalStateException(errorMsg);
            }
            
            if (depth > MAX_SERIALIZATION_DEPTH) {
                String errorMsg = String.format(
                        "Max serialization depth (%d) exceeded while serializing %s. " +
                        "This may indicate a circular reference or deeply nested structure. Entity: %s",
                        MAX_SERIALIZATION_DEPTH,
                        entity.getClass().getName(),
                        safeToString(entity));
                logger.logError(errorMsg);
                throw new IllegalStateException(errorMsg);
            }
            
            stack.add(entity);
            SERIALIZATION_DEPTH.set(depth + 1);
            
            var meta = MetaRepository.getInstance().getMeta((Class<T>)entity.getClass());
            if (meta == null) {
                throw new IllegalArgumentException("Unknown entity class: " + entity.getClass().getSimpleName());
            }

                logger.logTrace("Serializing entity: type={}, depth={}, stackSize={}", 
                        meta.getTypeName(), depth, stack.size());

            bytes.add(EntityType.CUSTOM.getCode());
            writeStr(meta.getTypeName(), bytes);

            var fieldBytes = new ArrayList<Byte>();

            var fieldCount = meta.getFields()
                    .stream()
                .filter(it -> it.get(entity) != null)
                .map(it -> {
                    writeStr(it.name(), fieldBytes);
                    var fieldValue = it.get(entity);
                        logger.logTrace("  Serializing field: name={}, valueType={}", 
                                it.name(), 
                                fieldValue != null ? fieldValue.getClass().getSimpleName() : "null");
                    serialize(fieldValue, fieldBytes);
                    return 0;
                }).count();
            
            writeInt(Long.valueOf(fieldCount).intValue(), bytes);
            bytes.addAll(fieldBytes);
        } finally {
            stack.remove(entity);
            SERIALIZATION_DEPTH.set(depth);
            if (depth == 0) {
                stack.clear();
            }
        }
    }
    
    private String safeToString(Object entity) {
        try {
            String str = entity.toString();
            if (str.length() > 200) {
                return str.substring(0, 200) + "...";
            }
            return str;
        } catch (Exception e) {
            return entity.getClass().getName() + "@" + System.identityHashCode(entity) + 
                    " (toString failed: " + e.getMessage() + ")";
        }
    }

    @Override public Object  deserializeEntity(Iterator<Byte> bytes) {
        var typeName = readStr(bytes);
        Meta<Object> meta = MetaRepository.getInstance().getMeta(typeName);
        if (meta == null) {
            throw new IllegalArgumentException("Unknown entity type-name: " + typeName);
        }
        var fieldCount = readInt(bytes);
        var fieldMap = new HashMap<String, Object>();

        for (int i  = 0; i  < fieldCount; i ++) {
            var fieldName = readStr(bytes);
            var fieldValue = deserialize(bytes);
            fieldMap.put(fieldName, fieldValue);
        }

        return meta.newInstance(fieldMap);
    }


    @Override public  boolean canTransform(Object entity) {
        throw new IllegalStateException("Unexpected call to CustomEntityTransform.canTransform");
    }

}

File: entity-meta-generator-lib/src/main/kotlin/com/loudsight/meta/serialization/transform/CustomEntityTransform.kt
package com.loudsight.meta.serialization.transform

import com.loudsight.meta.Meta
import com.loudsight.meta.MetaRepository
import com.loudsight.meta.serialization.EntityTransform
import com.loudsight.meta.serialization.EntityType

object CustomEntityTransform : EntityTransform<Any>(EntityType.CUSTOM, Any::class) {

    override fun serializeEntity(entity: Any, bytes: MutableList<Byte>) {
        serializeEntityX(entity, bytes)
    }

    private fun <T: Any> serializeEntityX(entity: T, bytes: MutableList<Byte>) {

        val meta = MetaRepository.getMeta(entity::class.java as Class<T>)
            ?: throw IllegalArgumentException("Unknown entity class: " + entity::class.simpleName)

        bytes.add(EntityType.CUSTOM.code)
        writeStr(meta.typeName, bytes)

        val fieldBytes = mutableListOf<Byte>()

//        val subtypeFields = meta.typeHierarchy
//            .map { MetaRepository.getMeta(it) as Meta<T> }
//            .flatMap { it.fields }
//
//        val fields = Stream.concat(subtypeFields.stream(), meta.fields.stream())
//            .collect(Collectors.toList())


        val fieldCount = meta.fields
            .filter { it[entity] != null }
            .map {
                writeStr(it.name, fieldBytes)
                val fieldValue = it[entity]!!
                serialize(fieldValue, fieldBytes)
            }.count()
        
        writeInt(fieldCount, bytes)
        bytes.addAll(fieldBytes)
    }

    override fun deserializeEntity(bytes: Iterator<Byte>): Any {
        val typeName = readStr(bytes)
        val meta: Meta<Any> = MetaRepository.getMeta(typeName)
            ?: throw IllegalArgumentException("Unknown entity type-name: $typeName")
        val fieldCount = readInt(bytes)
        val fieldMap = mutableMapOf<String, Any>()
//        fieldMap["__className__"] = typeName

        for (i in 0 until fieldCount) {
            val fieldName = readStr(bytes)
            val fieldValue = deserialize<Any>(bytes)
            fieldMap[fieldName] = fieldValue
        }

        return meta.newInstance(fieldMap)
    }


    override fun canTransform(entity: Any): Boolean {
        throw IllegalStateException("Unexpected call to CustomEntityTransform.canTransform")
    }

}
